notebooks/llmc_thunder.py

```diff
@@ -0,0 +1,1618 @@
+)
+# }
+
+
+# Attention-related {
+def _permute(qkv, n_embd, n_head, B, T, C):
+    q, k, v = qkv.split(n_embd, dim=2)
+    q = q.view(B, T, n_head, C // n_head).transpose(1, 2) # (B, nh, T, hs)
+    k = k.view(B, T, n_head, C // n_head).transpose(1, 2) # (B, nh, T, hs)
+    v = v.view(B, T, n_head, C // n_head).transpose(1, 2) # (B, nh, T, hs)
+    return q, k, v
+
+
+def _unpermute(y, B, T, C):
+    return y.transpose(1, 2).contiguous().view(B, T, C)
+
+
+def _manual_attention(qkv, bias, n_embd, n_head, B, T, C):
+    q, k, v = _permute(qkv, n_embd, n_head, B, T, C)
+
+    # manual implementation of attention
+    att = (q @ k.transpose(-2, -1)) * (1.0 / math.sqrt(k.size(-1)))
+    att = att.masked_fill(bias[:,:,:T,:T] == 0, float('-inf'))
+    att = F.softmax(att, dim=-1)
+    y = att @ v # (B, nh, T, T) x (B, nh, T, hs) -> (B, nh, T, hs)
+
+    y = _unpermute(y, B, T, C) # re-assemble all head outputs side by side
+
+    return y
+
+
+def llmc_permute_util_meta(qkv, n_embd, n_head, B, T, C):
+    shape = (B, n_head, T, C // n_head)
+    return (
+        TensorProxy(like=qkv, shape=shape),
+        TensorProxy(like=qkv, shape=shape),
+        TensorProxy(like=qkv, shape=shape),
+    )
+
+
+llmc_permute_util = llmc.register_operator(
+    "llmc_permute_util",
+    like=llmc_permute_util_meta,
+    fn=_permute,
+    replaces=_permute,
+)
+
+
+def llmc_permute_util_execution_transform(qkv, n_embd, n_head, B, T, C):
+    qkv = thunder.torch.reshape(qkv, (B, T, 3, n_head, C // n_head))
+    return llmc_permute(qkv)
+
+
+@langctx(Languages.TORCH)
+def llmc_permute_util_grad_transform(qkv, n_embd, n_head, B, T, C):
+    qkv_reshaped = thunder.torch.reshape(qkv, (B, T, 3, n_head, C // n_head))
+
+    q, k, v = llmc_permute(qkv_reshaped)
+
+    q_grad = get_grad(q)
+    k_grad = get_grad(k)
+    v_grad = get_grad(v)
+
+    grad = llmc_permute_backward(q_grad, k_grad, v_grad)
+    grad = thunder.torch.reshape(grad, qkv.shape)
+
+    put_grad(qkv, grad)
+
+    return q, k, v
+
+
+llmc.register_implementation(
+    llmc_permute_util,
+    checker=lambda *args, **kwargs: True,
+    execution_transform=llmc_permute_util_execution_transform,
+    grad_transform=llmc_permute_util_grad_transform,
+)
+
+
+def llmc_unpermute_util_meta(input: TensorProxy, B: int, T: int, C: int):
+    return TensorProxy(like=input, shape=(B, T, C))
+
+
+llmc_unpermute_util = llmc.register_operator(
+    "llmc_unpermute_util",
+    like=llmc_unpermute_util_meta,
+    fn=_unpermute,
+    replaces=_unpermute,
+)
+
+
+def llmc_unpermute_util_execution_transform(input: TensorProxy, B: int, T: int, C: int):
+    res = llmc_unpermute(input)
+    return thunder.torch.view(res, B, T, C)
+
+
+def llmc_unpermute_util_grad_transform(input: TensorProxy, B: int, T: int, C: int):
+    x = llmc_unpermute(input)
+    fwd = thunder.torch.view(x, B, T, C)
+
+    # NOTE: get_grad(x) breaks things. Why?
+    fwd_grad = get_grad(fwd)
+    fwd_grad = thunder.torch.view(fwd_grad, *x.shape)
```

Comment:
`get_grad(llmc_unpermute(input))` should be working and it's a bug that it doesn't.

Reply:
Because I need a non-reshaped intermediate.

Reply:
Yes, the documentation should be updated. Could you do that or create an issue about this?
Maybe it would be useful to get `get_grad` working on any intermediate, it doesn't work today because it's not an actual operation, it's a marker for the transform to indicate what are the inputs to the implicit backward function.